onnxruntime/core/providers/shared_library/provider_interfaces.h

```diff
@@ -0,0 +1,480 @@
+
+struct ProviderHost;
+struct Provider_IExecutionProvider;
+
+struct Provider_IExecutionProviderFactory {
+  virtual ~Provider_IExecutionProviderFactory() = default;
+  virtual std::unique_ptr<Provider_IExecutionProvider> CreateProvider() = 0;
+};
+
+//struct KernelCreateInfo;
+
+class DataTypeImpl;
+using MLDataType = const DataTypeImpl*;
+
+struct Provider_OrtDevice {
+  virtual ~Provider_OrtDevice() {}
+};
+
+struct Provider_OrtMemoryInfo {
+  static std::unique_ptr<Provider_OrtMemoryInfo> Create(const char* name_, OrtAllocatorType type_, Provider_OrtDevice* device_ = nullptr, int id_ = 0, OrtMemType mem_type_ = OrtMemTypeDefault);
+  virtual ~Provider_OrtMemoryInfo() {}
+
+  void operator=(const Provider_OrtMemoryInfo& v) = delete;
+};
+
+template <typename T>
+using Provider_IAllocatorUniquePtr = std::unique_ptr<T, std::function<void(T*)>>;
+
+struct Provider_IAllocator {
+  virtual ~Provider_IAllocator() {}
+
+  virtual void* Alloc(size_t size) = 0;
+  virtual void Free(void* p) = 0;
+
+  static bool CalcMemSizeForArray(size_t nmemb, size_t size, size_t* out) noexcept {
+    return CalcMemSizeForArrayWithAlignment<0>(nmemb, size, out);
+  }
+
+  template <size_t alignment>
+  static bool CalcMemSizeForArrayWithAlignment(size_t nmemb, size_t size, size_t* out) noexcept ORT_MUST_USE_RESULT;
+
+  template <typename T>
+  static Provider_IAllocatorUniquePtr<T> MakeUniquePtr(std::shared_ptr<Provider_IAllocator> allocator, size_t count_or_bytes) {
+    if (allocator == nullptr) return nullptr;
+    // for now limit to fundamental types. we could support others, but to do so either we or the caller
+    // needs to call the dtor for the objects, for buffers allocated on device we don't have destructor
+    //static_assert(std::is_fundamental<T>::value, "Fundamental type required as no destructors are called.");
+
+    size_t alloc_size = count_or_bytes;
+
+    // if T is not void, 'count_or_bytes' == number of items so allow for that
+    if (!std::is_void<T>::value) {
+      // sizeof(void) isn't valid, but the compiler isn't smart enough to ignore that this line isn't
+      // reachable if T is void. use std::conditional to 'use' void* in the sizeof call
+      if (!CalcMemSizeForArray(count_or_bytes, sizeof(typename std::conditional<std::is_void<T>::value, void*, T>::type),
+                               &alloc_size)) return nullptr;
+    }
+    return Provider_IAllocatorUniquePtr<T>{
+        static_cast<T*>(allocator->Alloc(alloc_size)),  // allocate
+        [=](T* ptr) { allocator->Free(ptr); }};         // capture IAllocator so it's always valid, and use as deleter
+  }
+
+  void operator=(const Provider_IAllocator& v) = delete;
+};
+
+template <size_t alignment>
+bool Provider_IAllocator::CalcMemSizeForArrayWithAlignment(size_t nmemb, size_t size, size_t* out) noexcept {
+  static constexpr size_t max_allowed = (static_cast<size_t>(1) << (static_cast<size_t>(std::numeric_limits<size_t>::digits >> 1))) - alignment;
```

Comment:
Does this need to be updated to copy the new implementation?

Reply:
Yes, did it just change today?

Reply:
Few weeks ago. 

---
In reply to: [419820633](https://github.com/microsoft/onnxruntime/pull/3314#discussion_r419820633) [](ancestors = 419820633)

Reply:
Updated and put a TODO to unify it in the future